index.d.ts

```diff
@@ -68,41 +104,27 @@ declare module "hyperapp" {
   // ---------------------------------------------------------------------------
 
   // A dispatched action handles an event in the context of the current state.
-  type Dispatch<S> = (action: Action<S>, payload?: Payload<any>) => void
+  type Dispatch<S> = (dispatchable: Dispatchable<S>, payload?: any) => void
+
+  // A dispatchable entity, when processed, causes a state transition.
+  type Dispatchable<S, P = any> =
+    | S
+    | StateWithEffects<S>
+    | Action<S, P>
+    | ActionWithPayload<S, P>
 
   // An action transforms existing state and/or wraps another action.
-  type Action<S, P = any> = ActionTransform<S, P> | ActionWithPayload<S, P>
-  type ActionTransform<S, P = any> = (
-    state: State<S>,
-    payload: Payload<P>
-  ) => State<S> | StateWithEffects<S> | Action<S>
-  type ActionWithPayload<S, P> = [ActionTransform<S, P>, Payload<P>]
-
-  // A transform carries out the transition from one state to another.
-  type Transform<S, P = any> = (
-    state: State<S> | StateWithEffects<S>,
-    payload?: Payload<P>
-  ) => State<S> | StateWithEffects<S>
-
-  // Application state is accessible in every view, action, and subscription.
-  type State<S> = S
+  type Action<S, P = any> = (state: S, payload: P) => Dispatchable<S>
+  type ActionWithPayload<S, P = any> = [action: Action<S, P>, payload: P]
 
   // State can be associated with a list of effects to run.
-  type StateWithEffects<S, D = any> = [State<S>, ...Effect<S, D>[]]
-
-  type EffectCreator<S, D = any> = (..._: any[]) => Effect<S, D>
+  type StateWithEffects<S, P = any> = [state: S, ...effects: Effect<S, P>[]]
 
   // An effect is an abstraction over an impure process.
-  type Effect<S, D = any> = [Effecter<S, D>, Payload<D>]
+  type Effect<S, P = any> = [effecter: Effecter<S, P>, payload: P]
 
   // An effecter is where side effects and any additional dispatching may occur.
-  type Effecter<S, D> = (
-    dispatch: Dispatch<S>,
-    payload?: Payload<D>
-  ) => void | Promise<void>
-
-  // A payload is data given to an action, effect, or subscription.
-  type Payload<P> = P
+  type Effecter<S, P> = (dispatch: Dispatch<S>, payload: P) => void | Promise<void>
```

Comment:
Let's remove the `| Promise<void>` for now. I don't know if I'll want to support this right now.

Reply:
Without `| Promise<void>` using stuff `fetch()` is impossible.  I'm using async effects to implement an app-specific API I have.

Reply:
How would you do async code in a Hyperapp app?  Delegate that stuff to an external layer of custom code outside of the  Hyperapp app ?

Reply:
You can use `fetch()`, what you can't do is declare the function as `async`, that's all.

Reply:
You can use `fetch()`, what you won't be able to do is declare the function as `async`, that's it.

Reply:
Umm... use `fetch()` _where exactly_?  If I press a button in my app and that triggers a fetch to some resource, what does that code look like if effecters cannot return `Promise<void>`?

Reply:
A function doesn't need to be declared as `async` for you to use `fetch()`. You just use `fetch().then(...)`. 

We could definitely allow `async` effect implementations in the future, but don't have to right now.

Reply:
Let's say for the sake of argument what you say is correct, that using `fetch().then(...)` is an alternative.

Well, then you're basically dictating to the user how they should do their async coding which I believe is outside of Hyperapp's purview.

And for what gain, exactly ?

Reply:
If effecters and subscribers are sanctioned areas where the user can do their messy imperative stuff, then let them do it the way they want.

Reply:
That's a great point, which is exactly why I think we can consider this again in the future.

Reply:
Could you clarify? Which one to consider in the future?

Reply:
Allowing `async`-`await` effect implementations.

Reply:
The types already allow that.  What is the benefit to removing this now when it's not known to cause any issues while allowing a modern style of async coding ?

Reply:
I think Promise<void> needs to stay on the basis that anything that is possible in plain JS should be possible in TS (at least if it adds some value, however slight). 

Due to the interconnected nature of the types, it won’t be possible to add your own version of Effect that supports this, without redefining all the types. So I guess technically it will be possible but in a quite obnoxious way.

I also don’t prefer using async/await over raw Promises, but it is popular (especially among the TS crowd it seems). It won’t be long before people beg to have it put in (or are alienated for the lack)

Reply:
> I think Promise needs to stay on the basis that anything that is possible in plain JS should be possible in TS (at least if it adds some value, however slight).

Hmm, I see what you mean. What's already possible in JS should not be made impossible in TS. Got it. 

In this case, the types just present the world the way it is. Hyperapp is not necessarily advocating for `async` effects, but no one can stop you from using `async`/`await` if that's what you want.

Reply:
> the types just present the world the way it is
☝️ exactly!

Reply:
Actually... if your issue is with using the word `Promise` in our type defs @jorgebucaran, we could just have:

```js
type Effecter<S, P> = (dispatch: Dispatch<S>, payload: P) => any
```

... and in fact maybe we _should_ have that. Because hyperapp doesn't care what an effecter returns, right?

Reply:
Makes sense too! More, actually.

Reply:
Interesting, so Hyperapp simply does nothing with the return type, right ?

Reply:
Same goes for subscribers I'd imagine ?

Reply:
Subscribers are supposed to return unsubscribers. But yeah, Hyperapp ignores the returns value of unsubscribers.

Reply:
Oh, right, duh !

Reply:
Carrying on this discussion here: https://github.com/jorgebucaran/hyperapp/discussions/1052

This particular topic is at the top of the list.